### PR TITLE
Add Softmax

### PR DESCRIPTION
**Pull request by rhemon**

### CODE REVIEW

**Comment by rhemon** [comment]
12 - 4
0.02 learning rate
3000 iterations
```
Accuracy with training set, 85.890625%
Accuracy with dev set, 83.75%
Accuracy in recognising mango: 79.1111111111
Accuracy in recognising apple: 90.1111111111
Accuracy in recognising orange: 94.3333333333
Accuracy in recognising pear: 69.7777777778
```
[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/84)

**Comment by rhemon** [comment]
# 2 Layer Neural Network [12, 4]

## 10000 Iterations

### 0.00005 learning rate

```
Accuracy with training set, 33.90625%
Accuracy with dev set, 43.5%
```
[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/185)

### 0.00075 learning rate

```
Accuracy with training set, 48.09375%
Accuracy with dev set, 53.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/186)

### 0.005 learning rate

```
Accuracy with training set, 69.78125%
Accuracy with dev set, 62.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/187)

### 0.01 learning rate

```
Accuracy with training set, 79.34375%
Accuracy with dev set, 61.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/188)

## 15000 iterations

### 0.00005 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/190)

### 0.00075 learning rate

```
Accuracy with training set, 49.34375%
Accuracy with dev set, 53.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/191)

### 0.005 learning rate

```
Accuracy with training set, 75.59375%
Accuracy with dev set, 61.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/192)

### 0.01 learning rate

```
Accuracy with training set, 83.40625%
Accuracy with dev set, 59.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/193)

## 20000 iterations

### 0.00005 learning rate

```
Accuracy with training set, 36.09375%
Accuracy with dev set, 42.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/195)

### 0.00075 learning rate

```
Accuracy with training set, 49.71875%
Accuracy with dev set, 54.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/196)

### 0.005 learning rate

```
Accuracy with training set, 79.15625%
Accuracy with dev set, 62.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/197)

### 0.01 learning rate

```
Accuracy with training set, 87.09375%
Accuracy with dev set, 59.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/198)

## 25000 iterations

### 0.00005 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/200)

### 0.00075 learning rate

```
Accuracy with training set, 51.53125%
Accuracy with dev set, 56.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/201)

### 0.005 learning rate

```
Accuracy with training set, 82.65625%
Accuracy with dev set, 59.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/202)

### 0.01 learning rate

```
Accuracy with training set, 89.53125%
Accuracy with dev set, 59.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/203)

## 30000 iterations

### 0.00005 learning rate

```
Accuracy with training set, 33.71875%
Accuracy with dev set, 44.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/205)

### 0.00075 learning rate

```
Accuracy with training set, 54.28125%
Accuracy with dev set, 58.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/206)

### 0.005 learning rate

```
Accuracy with training set, 84.15625%
Accuracy with dev set, 62.5%
```

[CircleC Log](https://circleci.com/gh/rhemon/food_rec/207)

### 0.01 learning rate

```
Accuracy with training set, 92.46875%
Accuracy with dev set, 61.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/208)

## 40000 iterations

### 0.005 learning rate

```
Accuracy with training set, 87.59375%
Accuracy with dev set, 60.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/210)

### 0.01 learning rate

```
Accuracy with training set, 94.53125%
Accuracy with dev set, 60.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/211)

**Comment by rhemon** [comment]
# 3 Layer Neural Network [24, 12, 4]

## 10000 Iterations

### 0.00005 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```
[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/159)

### 0.00075 learning rate

```
Accuracy with training set, 32.15625%
Accuracy with dev set, 42.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/160)

### 0.005 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/161)

### 0.01 learning rate

```
Accuracy with training set, 38.21875%
Accuracy with dev set, 44.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/162)

### 0.05 learning rate

```
Accuracy with training set, 49.59375%
Accuracy with dev set, 52.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/163)

## 15000 iterations

### 0.00005 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/164)

### 0.00075 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/165)

### 0.005 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/166)

### 0.01 learning rate

```
Accuracy with training set, 38.34375%
Accuracy with dev set, 43.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/167)

### 0.05 learning rate

```
Accuracy with training set, 82.09375%
Accuracy with dev set, 62.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/168)

## 20000 iterations

### 0.00005 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/169)

### 0.00075 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/171)

### 0.005 learning rate

```
Accuracy with training set, 38.03125%
Accuracy with dev set, 44.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/172)

### 0.01 learning rate

```
Accuracy with training set, 39.78125%
Accuracy with dev set, 45.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/173)

### 0.05 learning rate

```
Accuracy with training set, 92.34375%
Accuracy with dev set, 64.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/174)

## 25000 iterations

### 0.00005 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/175)

### 0.00075 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/176)

### 0.005 learning rate

```
Accuracy with training set, 38.53125%
Accuracy with dev set, 43.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/177)

### 0.01 learning rate

```
Accuracy with training set, 39.15625%
Accuracy with dev set, 44.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/178)

### 0.05 learning rate

```
Accuracy with training set, 93.71875%
Accuracy with dev set, 64.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/179)

## 30000 iterations

### 0.00005 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/180)

### 0.00075 learning rate

```
Accuracy with training set, 28.65625%
Accuracy with dev set, 37.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/181)

### 0.005 learning rate

```
Accuracy with training set, 41.28125%
Accuracy with dev set, 44.0%
```

[CircleC Log](https://circleci.com/gh/rhemon/food_rec/182)

### 0.01 learning rate

```
Accuracy with training set, 42.59375%
Accuracy with dev set, 42.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/183)

### 0.05 learning rate

```
Accuracy with training set, 94.96875%
Accuracy with dev set, 65.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/184)

<strong>Note: Models with learning rate 0.05 shows to have higher accuracy but cost at many turns increased and then dropped back.</strong>

**Comment by rhemon** [comment]
# 3 Layer [24,12,4]

## 0.01500 learning rate - 40000 iterations

```
Accuracy with training set, 72.84375%
Accuracy with dev set, 57.5%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/213)

## 0.02000 learning rate - 40000 iterations

```
Accuracy with training set, 87.15625%
Accuracy with dev set, 63.0%
```

<strong> Cost started going up down after around 36600 turns </strong>

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/214)

## 0.02500 learning rate - 40000 iterations

```
Accuracy with training set, 92.96875%
Accuracy with dev set, 62.5%
```

<strong> Cost started going up down after around 30300 turns </strong>

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/215)

## 0.01000 learning rate - 50000 iterations

```
Accuracy with training set, 52.15625%
Accuracy with dev set, 39.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/217)


## 0.01500 learning rate - 50000 iterations

```
Accuracy with training set, 87.71875%
Accuracy with dev set, 62.0%
```

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/218)

## 0.02000 learning rate - 50000 iterations

```
Accuracy with training set, 93.84375%
Accuracy with dev set, 64.5%
```

<strong> Cost goes up and down </strong>

[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/219)

Learning rates >= 0.0200 does lead to higher accuracy in training set but the cost at turns sometimes increase. 
Also learning rates <= 0.0100 seems like it hardly affects, even with lots of iterations, on the other hand with the same learning rate, 2 layer network performs a lot better with 30000 iterations.

**Comment by rhemon** [comment]
## 2 layer, 0.01 learning rate, 80000 iterations
```
Accuracy with training set, 96.03125%
Accuracy with dev set, 61.0%
```
[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/224)

## 3 layer, 0.015 learning rate, 80000 iterations
```
Accuracy with training set, 95.21875%
Accuracy with dev set, 64.0%
```
[CircleCI Log](https://circleci.com/gh/rhemon/food_rec/222)
